Spar hints for operations. (#871)
wireapp/wire-server: docs/reference/spar-braindump.md

# the spar service for user provisioning (scim) and authentication (saml) - a brain dump

this is a mix of information on inmplementation details, architecture,
and operation.  it should probably be sorted into different places in
the future, but if you can't find any more well-structured
documentation answering your questions, look here!

## operations

> ### enabeling / disabeling the sso feature for a team

Reviewer: ```suggestion
### enabling / disabling the sso feature for a team
```

if you have sso disabled by default, you need to turn on the feature
for every team that wants to use it.  you can do this in the stern
service (aka backoffice).  look for `get/put
/teams/{tid}/features/sso`

### registering an idp with a team via curl

you need to have:

```sh
# user id of an admin of the team (or the creator from the team info
# in the backoffice, if you only have the team id).
export ADMIN_ID=...

# path of the xml metadata file (if you only have the url, curl it)
export METADATA_FILE=...
```

then you need to do:

```sh
./khan.me scp upload -e prod -r spar -s .../wire-server/deploy/services-demo/register_idp_internal.sh -d ./register.sh
./khan.me scp upload -e prod -r spar -s ${METADATA_FILE} -d ./metadata.xml
./khan.me ssh -e prod -r spar
export TEAM_OWNER_ID=...
./register.sh metadata.xml $TEAM_OWNER_ID
```

> the output contains the a json object representing the idp.  safe the `id` field of that object.

Reviewer: ```suggestion
the output contains the a json object representing the idp.  Save the `id` field of that object.
```

Reviewer: Save it where? Why?

the sso login code needed by the users to authenticate via saml is

```
wire-<that-id>
```

> ### trouble shooting

Reviewer: ```suggestion
### troubleshooting
```

- ask user for screenshots of the error message, or even better, for
  the text.  the error message contains lots of strings that you can
  grep for in the spar sources.

#### making spar work with a new IdP

often, new IdPs work out of the box, because there appears to be some
consensus about what minimum feature set everybody should support.

if there are problems: collect the metadata xml and an authentication
response xml (either from the browser http logs via a more technically
savvy customer; FUTUREWORK: it would be nice to log all saml response
xml files that spar receives in prod and cannot process).

https://github.com/wireapp/saml2-web-sso supports writing [unit vendor
compatibility
tests](https://github.com/wireapp/saml2-web-sso/blob/ff9b9f445475809d1fa31ef7f2932caa0ed31613/test/Test/SAML2/WebSSO/APISpec.hs#L266-L329)
against that response value.  once that test passes, it should all
work fine.

### common misconceptions

#### an email address can be one of two things

the email used for saml auth is only a name, and never used for
sending out emails, and does not show as the email address of the user
in the team settings.

> RATIONALE: emails that are passed in from an external identity

Reviewer: I dont fully understand what you mean. Why cant we send emails to trusted email addresses?

Author: we can, but we haven't gotten around to actually doing it yet.  there is a FUTUREWORK right below that rationale, i think.  what should i do to make this clearer?

Reviewer: I will rewrite this in my own words. I understand it now.

provider must be trusted, so the user cannot have them as an actual
email address that wire is sending emails to.

POSSIBLE FEATURE: we could authenticate the emails sent in from the
identity provider in the same way we are doing that for
password-authenticated non-team users: email receives a link
containing a crypto token, user clicks on link if the email is
authentic, email gets authenticated.

#### scim, provisioning, metadata

> for changing the user information (name, handle, email, ...), saml

Reviewer: do you have some overview of what SSO providers actually provide scim?

Author: i don't.  i know that azure, octa have it, and i *think* centrify as well.  that's the one we have worked with the most.  it's also not a big deal to hack it together in a shell script, so any software really can support it with a little love from a good ops person.  example: https://github.com/wireapp/wire-server/blob/develop/docs/reference/provisioning/scim-via-curl.md

nor sure: should i move this comment into the actual markdown?  fully or partially?

Reviewer: yes. having an incomplete list of what providers we actually support sounds useful.

Author: 0cd2f519

other places are allowed, and must be checked by the application code
for danglingness.  ([test case for
scim](https://github.com/wireapp/wire-server/blob/010ca7e460d13160b465de24dd3982a397f94c16/services/spar/test-integration/Test/Spar/Scim/UserSpec.hs#L239-L308);
[test case for
saml](https://github.com/wireapp/wire-server/blob/293518655d7bae60fbcb0c4aaa06034785bfb6fc/services/spar/test-integration/Test/Spar/APISpec.hs#L742-L795))

For the semantics of interesting corner cases, consult [the test
suite](https://github.com/wireapp/wire-server/blob/develop/services/spar/test-integration/Test/Spar/APISpec.hs).
If you can't find what you're looking for there, please add at least a
pending test case explaining what's missing.

Side note: Users in brig carry an enum type
[`ManagedBy`](https://github.com/wireapp/wire-server/blob/010ca7e460d13160b465de24dd3982a397f94c16/libs/brig-types/src/Brig/Types/Common.hs#L393-L413);
see also {#DevScimOneWaySync}.  This is a half-implemented feature for
managing conflicts between changes via scim vs. changes from wire
clients; and does currently not affect deletability of users.  (grep

> `wire-server` for up-to-date information.)

Reviewer: What do you mean here with up to date information?

Author: i mean "it is conceivable that this documentation is out of date, please consult the source".

i'm emberrassed now.  remove that bracket?
